File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.go
import "math"

func maxSubarrayLength(nums []int, k int) int {
    n := len(nums)
    hash := make(map[int]int)
    l, r, ans := 0, 0, math.MinInt

    for ; r < n; r++ {
        hash[nums[r]]++

        for hash[nums[r]] > k {
            hash[nums[l]]--
            l++
        }
        ans = max(ans, r - l + 1)
    }
    return ans
}

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cs
public class Solution {
    public int MaxSubarrayLength(int[] nums, int k) {
        int n = nums.Length;
        var map = new Dictionary<int, int>();
        int l = 0, r = 0, ans = Int32.MinValue;

        for (; r < n; r++) {
            if (map.ContainsKey(nums[r]))
                map[nums[r]]++;
            else
                map[nums[r]] = 1;
            
            while (map[nums[r]] > k)
                map[nums[l++]]--;
            
            ans = Math.Max(ans, r - l + 1);
        }
        return ans;
    }
}